includes/constants/class-base-constant.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Class Base_Constant
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Constants;
+
+use ReflectionClass;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+
+/**
+ * Base constant class to hold common logic for all constants.
+ */
+abstract class Base_Constant {
+
+	/**
+	 * Enum value
+	 *
+	 * @var mixed
+	 */
+	protected $value;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param mixed $value Constant from class.
+	 */
+	public function __construct( $value ) {
+		if ( $value instanceof static ) {
+			$value = $value->get_value();
+		}
+		$this->value = $value;
```

Comment:
I think that a better option would be to move the constant validation form _callStatic to the constructor, so we don't have to use the search function. The search function is used to get a constant name from a constant value, so in this case it will not be good option to use.